plugins/experimental/traffic_dump/session_data.h

```diff
@@ -61,9 +62,12 @@ class SessionData
   /// Whether the first transaction in this session has been written.
   bool has_written_first_transaction = false;
 
-  TSCont aio_cont       = nullptr; /// AIO continuation callback
-  TSCont txn_cont       = nullptr; /// Transaction continuation callback
-  TSMutex disk_io_mutex = nullptr; /// AIO mutex
+  TSCont aio_cont = nullptr; /// AIO continuation callback
+  TSCont txn_cont = nullptr; /// Transaction continuation callback
+
+  // The following has to be recursive because the stack does not unwind
+  // between event invocations.
+  std::recursive_mutex disk_io_mutex; /// The mutex for guarding IO calls.
```

Comment:
I think you might be able to avoid making the mutex recursive.  You'd have to call the reenable function in a lambda function that you passed to the constructor of an instance of ts::PostScript (created before the lock_guard).  But it's probably not worth the bother.

Reply:
https://github.com/apache/trafficserver/blob/0b92e7cc3a6df779c4d527e0915adfb7ed936487/plugins/xdebug/xdebug.cc#L533